Adds the mounted Leach-Rifle to replace the laser gun implant

itseasytosee opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
So this PR adds the mounted Leach-rifle as a replacement for the current counted laser gun implant that we have but before that, I want to make some points about what was wrong with the laser gun

- Firstly, it was scarcely used due to the fact that it was role restricted to a role that already has a lot of potential for different more appealing traitorous acts such as emaging borgs or utilizing a concealed weapon bay. The laser gun implant doesn't connect with the roboticist in any way except for its theming.
- Secondly, it is an uninteresting consept. It was just a captain's laser gun glued into your arm. This also makes the captain's laser gun feel less special.
- Thirdly, and this is more personal, I dislike going against the established "lore" that syndies use bullets and nanotrasen uses lasers

So what's changed?
The leach rifle fires bolts of energy that deal brute damage and can't pass through windows. It is no longer role restricted tc and it may need a tc cost change because of that. But ill get on to the main feature, the recharge.

The leach-gun recharges by sapping its wielder's energy leaving them tired and disorientated. after it is activated by using it in-hand it will start the recharge. The more charge is missing the longer the recharge process will take and the more stamina damage and dizziness its wielder will suffer. At the end of the recharge if the wielder's joint stamina brute and burn damage is over a threshold the wielder will be knocked down and stunned for a bit. Recharging is quite obvious; Sparks will fly and the wielder will twitch and scream.


Also, a minor refractor of mounted weapons so they're a bit simpler to add in the future.


## Why It's Good For The Game

Turns a pretty boring weapon into one with a neat mechanic. Not saying the laser gun is objectively bad, I just don't think it makes for a very good syndie weapon.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: itseasytosee
del: The Laser Arm Implant has been removed
add: The new Mounted Leach-Rifle has been added in its place.
/:cl:

itseasytosee committed:
leach gun

Fikou commented on code/game/objects/effects/temporary_visuals/miscellaneous.dm:
```diff
@@ -364,6 +364,10 @@
 	duration = 4
 
+/obj/effect/temp_visual/impact_effect/yellow_laser
+	icon = 'icons/obj/projectiles.dmi'
```
? this is either inherited from parent or you put that laser icon in the wrong file

itseasytosee replied:
Yeah ill fix it I was just lazy

Fikou commented on code/modules/projectiles/guns/energy/mounted.dm:
```diff
@@ -1,26 +1,78 @@
-	trigger_guard = TRIGGER_GUARD_ALLOW_ALL // Has no trigger at all, uses neural signals instead
 
-/obj/item/gun/energy/e_gun/advtaser/mounted/dropped()//if somebody manages to drop this somehow...
+/obj/item/gun/energy/mounted/dropped()
 	..()
```
this line is useless

itseasytosee replied:
It sure is!

Fikou commented on code/modules/projectiles/guns/energy/mounted.dm:
```diff
@@ -1,26 +1,78 @@
+		to_chat(user, "<span class='userwarning'>The [src] is already fully charged.</span>")
+		return
+	cooldown = (world.time + 20 SECONDS)
+	to_chat(user, "<span class='userdanger'>You begin the recharging proccess...</span>")
```
does this really need userdanger? it doesnt do any actual damage but youll have giant red messages

Fikou commented on code/modules/projectiles/guns/energy/mounted.dm:
```diff
@@ -1,26 +1,78 @@
+	to_chat(user, "<span class='userdanger'>You begin the recharging proccess...</span>")
+	do
+		if(do_after(user, 3 SECONDS, timed_action_flags = (IGNORE_USER_LOC_CHANGE|IGNORE_HELD_ITEM)))
+			to_chat(user, "<span class='userdanger'>[pick("It hurts...", "You feel your energy being draining away...", "Your body goes numb...")]</span>")
```
same here

ATH1909 replied:
"being draining away"

Fikou commented on code/modules/projectiles/guns/energy/mounted.dm:
```diff
@@ -1,26 +1,78 @@
+			return
+	while(src.cell.percent() < 100)
+	if(do_after(user, 7 SECONDS, timed_action_flags = (IGNORE_USER_LOC_CHANGE|IGNORE_HELD_ITEM)))
+		to_chat(user, "<span class='userdanger'>[pick("You feel relief as the draining stops...", "The pain is over...", "The strain on your body subsides...")]</span>")
```
and here

Fikou commented on code/modules/projectiles/projectile/beams.dm:
```diff
@@ -60,6 +60,16 @@
 
+/obj/projectile/beam/leach
+	name = "leach shot"
+	damage = 18
```
for 7 tc i can buy a makarov pistol, concealable in a box, with a lot of ammo, without needing to spend a lot of time and stamina on recharge, and it will do 30 damage per shot

necromanceranne replied:
For a skirmish weapon, a 12 tc ebow is usually a better choice with no strings attached to the leach rifle as well. Infinite ammo with a delay between shots you can work around.

It's okay to make antags harder to actually engage constantly, since that's what they're going to be doing if they need something like this to keep skirmishing and fighting for whatever reason.

I did say this I think when the original pr to add the laser implant came up, but I feel like this is another redundant weapon which already has items filling the same space.

ATH1909 commented on code/modules/projectiles/guns/energy/mounted.dm:
```diff
@@ -1,26 +1,78 @@
+	if(src.cell.percent() == 100)
+		to_chat(user, "<span class='userwarning'>The [src] is already fully charged.</span>")
+		return
+	cooldown = (world.time + 20 SECONDS)
```
does this really need such a long cooldown? what if someone bumps you during the do_after?

itseasytosee replied:
if the gun is fully spent it will take 19 seconds to recharge. 3 seconds per 30% missing charge and then 7 seconds once the process is over. this is to stop multiple charges to be queued up

itseasytosee replied:
ill clean this up and remove the cooldown

ATH1909 commented on code/modules/projectiles/guns/energy/mounted.dm:
```diff
@@ -1,26 +1,78 @@
+		else
+			emergencydump(user)
+			return
+	while(src.cell.percent() < 100)
+	if(do_after(user, 7 SECONDS, timed_action_flags = (IGNORE_USER_LOC_CHANGE|IGNORE_HELD_ITEM)))
```
this indentation looks wonky

ATH1909 commented on code/modules/projectiles/guns/energy/mounted.dm:
```diff
@@ -1,26 +1,78 @@
+
+/obj/item/gun/energy/mounted/leachgun/proc/emergencydump(mob/living/user) // used when somthing unexpected happens durring the recharge
+		to_chat(user,"<span class='warning'>UNEXPECTED ANOMALY: Emergency dropout protocol initiated</span>")
+		user.adjustStaminaLoss(200, 0)
```
![image](https://user-images.githubusercontent.com/42606352/105570949-f0c1ec00-5d11-11eb-8ba8-d41b39553364.png)

necromanceranne replied:
Oh fuck it does 200 stamina damage if the drain is interrupted. Doesn't that just crush all your limbs?

itseasytosee replied:
that would be super funny but no it just stam-crits you

itseasytosee commented:
Thanks for your feedback i'll work on it

itseasytosee committed:
fixes

itseasytosee commented:
Ok, I've cleaned up the code and got rid of the user danger. I've also added a 10 stamina damage on every hot to make it viable against armored targets. If anyone had any suggestions on how to make this more unique or specialized I'd love to hear them.

SmArtKar commented:
Maybe small amounts of toxins as well would fit better than just stam damage only? Also, I really suggest making it deal more damage, like 20-25 maybe?

itseasytosee commented:
I'm not into adding toxin damage for this weapon, it doesn't make sense thematically. 20 damage a hit makes it a five-shot kill on unarmored targets at full health which I think is acceptable. 25 seems like too much considering how much leniency you get with the clip size and how easy it is to reload given a bit of downtime.

itseasytosee committed:
+2 dmg

necromanceranne commented:
10 stamina damage does nothing at all. I'm a little confused by its presence. Usually when you have damage mixes, it favours stamina damage over lethal due to lethal being sticky but stamina resets. Stamina and health are tracked separately for slowdown as well as overall critical state.

The inclusion of stamina damage is a moot point.

itseasytosee commented:
I've kind of realized this PR is pointless. I was justifying it as an improvement over what we currently have, but it's really not worth it. I'm closing tomorrow unless someone is really interested in it.

necromanceranne commented:
> I've kind of realized this PR is pointless. I was justifying it as an improvement over what we currently have, but it's really not worth it. I'm closing tomorrow unless someone is really interested in it.

The goal was noble but you're working with trash. It's not easy trying to justify bloat a second time.

itseasytosee commented:
this is so sad
